src/memtable.rs

```diff
@@ -1,27 +1,70 @@
+use crate::util::Comparator;
+use crate::value::Value;
+use crate::wal::Wal;
+use crate::AgateOptions;
+use crate::Result;
 use bytes::Bytes;
-use skiplist::{FixedLengthSuffixComparator as Flsc, Skiplist};
+use skiplist::Skiplist;
 use std::collections::VecDeque;
 use std::mem::{self, ManuallyDrop, MaybeUninit};
+
 use std::ptr;
+use std::sync::RwLock;
 
-pub struct MemTableView {
-    tables: ManuallyDrop<[Skiplist<Flsc>; 20]>,
-    len: usize,
+const MEMTABLE_VIEW_MAX: usize = 20;
+
+/// MemTableCore guards WAL and max_version.
+/// These data will only be modified on memtable put.
+/// Therefore, separating wal and max_version enables
+/// concurrent read/write of MemTable.
+struct MemTableCore {
+    wal: Option<Wal>,
+    max_version: u64,
 }
 
-impl MemTableView {
-    pub fn get(&self, key: &[u8]) -> Option<&Bytes> {
-        for i in 0..self.len {
-            let res = self.tables[i].get(key);
-            if res.is_some() {
-                return res;
-            }
+pub struct MemTable {
+    pub(crate) skl: Skiplist<Comparator>,
+    opt: AgateOptions,
+    core: RwLock<MemTableCore>,
```

Comment:
Why use `RwLock`?

Reply:
This `RwLock` protects `Wal`. It seems that when database is running, we only write to `Wal`, and we don't need to read from it when reading from memtable. Perhaps `Mutex` is a better choice, I'll change that.